services/headless-lms/server/src/controllers/main_frontend/exercise_repositories.rs

```diff
@@ -0,0 +1,191 @@
+use std::ops::Deref;
+
+use models::{
+    exercise_repositories::{ExerciseRepository, ExerciseRepositoryUpdate},
+    CourseOrExamId,
+};
+
+use crate::{controllers::prelude::*, domain};
+
+#[derive(Debug, Deserialize)]
+#[cfg_attr(feature = "ts_rs", derive(TS))]
+pub struct NewExerciseRepository {
+    course_id: Option<Uuid>,
+    exam_id: Option<Uuid>,
+    git_url: String,
+    deploy_key: Option<String>,
+}
+
+/**
+POST `/api/v0/main-frontend/exercise-repositories/new
+*/
+#[generated_doc]
+#[instrument(skip(pool, file_store, app_conf))]
+async fn new(
+    pool: web::Data<PgPool>,
+    file_store: web::Data<dyn FileStore>,
+    repository: web::Json<NewExerciseRepository>,
+    user: AuthUser,
+    app_conf: web::Data<ApplicationConfiguration>,
+) -> ControllerResult<web::Json<Uuid>> {
+    let mut conn = pool.acquire().await?;
+    let course_or_exam_id = CourseOrExamId::from(repository.course_id, repository.exam_id)?;
+    let token = authorize(
+        &mut conn,
+        Act::Edit,
+        Some(user.id),
+        Res::from_course_or_exam_id(course_or_exam_id),
+    )
+    .await?;
+    // create pending repository
+    let new_repository_id = Uuid::new_v4();
+    models::exercise_repositories::new(
+        &mut conn,
+        new_repository_id,
+        course_or_exam_id,
+        &repository.git_url,
+        repository.deploy_key.as_deref(),
+    )
+    .await?;
+    // processing a repository may take a while, so this is done in the background
+    actix_web::rt::spawn(async move {
```

Comment:
Maybe we should do this kind of background job in its own binary that would get scheduled as an independent kubernetes pod?